crates/common_lang_types/src/text_with_carats.rs

```diff
@@ -116,19 +114,35 @@ fn text_with_carats_and_line_count_buffer_and_line_numbers(
         };
 
         if should_print_carats {
-            let start_of_carats = actual_span.start.saturating_sub(start_of_line as u32);
             let line_len = line_content.len() as u32;
-            let end_of_carats = std::cmp::min(
+            let start_of_carats = actual_span.start.saturating_sub(start_of_line as u32);
+            let start_idx = start_of_carats as usize;
+            let end_idx = std::cmp::min(
                 actual_span.end.saturating_sub(start_of_line as u32),
                 line_len,
-            );
+            ) as usize;
+
+            let prefix = &line_content[0..start_idx];
+            let highlighted = &line_content[start_idx..end_idx];
+            let suffix = &line_content[end_idx..];
 
+            let colored_source = format!(
+                "{}{}{}",
+                prefix,
+                if colorize_carats {
+                    highlighted.bright_red().to_string()
+                } else {
+                    highlighted.to_string()
+                },
+                suffix
+            );
+            output_lines.push(colored_source);
             // a line may be entirely empty, due to containing only a \n. We probably want to avoid
             // printing an empty line underneath. This is weird and probably buggy!
+            let end_of_carats = end_idx as u32;
```

Comment:
instead of having two variables (`end_of_carats` and `end_idx`) which have the same value but different representations, can we have a single one, and just cast it as needed?

e.g. if `start_of_carats` became `let start_of_carats = (actual_span.start as usize).saturating_sub(start_of_line);` we probably could do with fewer variables and less casting.

I'm not concerned with the perf implications of usize vs u32 here, I just want this code to be understandable later

(and despite my previous comment about multiple PRs, feel free to make this change in this same PR)